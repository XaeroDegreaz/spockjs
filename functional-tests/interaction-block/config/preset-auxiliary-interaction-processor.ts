import { InteractionProcessor } from '@spockjs/config';

export const assertionPostProcessors = [];

export const interactionProcessors: InteractionProcessor[] = [
  t => ({
    primary: false,
    declare: () => t.emptyStatement(),
    verify: () => t.emptyStatement(),
  }),
];
