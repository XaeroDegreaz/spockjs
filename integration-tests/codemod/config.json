{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert"
      }
    ]
  ]
}
