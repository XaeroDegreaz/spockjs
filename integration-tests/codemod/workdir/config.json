{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert",
        "presets": ["@spockjs/preset-sinon-mocks"]
      }
    ]
  ]
}
