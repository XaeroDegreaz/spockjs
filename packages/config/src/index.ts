import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

import { Hooks } from './hooks';

/**
 * Config as it can be provided by the user.
 * Comments indicate transformations made to conform to InternalConfig format.
 */
export const configType = t.exact(
  t.partial({
    powerAssert: t.boolean,
    autoImport: t.union([t.boolean, t.string]),
    staticTruthCheck: t.boolean,
    assertFunctionName: t.string,
    presets: t.array(t.string),
  }),
  'Config',
);
export type Config = t.TypeOf<typeof configType>;

/**
 * Config that disables all but the core set of features.
 */
export const minimalConfig: Config = {
  powerAssert: false,
  autoImport: false,
};

/**
 * Config containing the settings used if they are missing in the user-specified config.
 */
export const defaultConfig = {
  powerAssert: true,
  autoImport: true,
  staticTruthCheck: false,
  assertFunctionName: '',
  presets: [],
};

/**
 * Config after sanitizing in extractConfigFromState.
 */
export interface InternalConfig {
  readonly powerAssert: boolean;
  readonly autoImport: string; // empty => no auto import
  readonly staticTruthCheck: boolean;
  readonly assertFunctionName: string; // empty => no custom assert function name
  readonly hooks: Hooks;
}

export const extractConfigFromState = ({
  opts,
}: {
  opts: any;
}): InternalConfig => {
  const decoded = configType.decode(opts);

  let {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    presets,
  } = decoded.getOrElseL(errors => {
    throw new Error(
      ['Invalid plugin configuration:', ...failure(errors)].join('\n'),
    );
  });

  // powerAssert
  if (powerAssert === undefined) {
    ({ powerAssert } = defaultConfig);
  }

  // autoImport
  if (autoImport === undefined) {
    ({ autoImport } = defaultConfig);
  }
  if (autoImport === false) {
    autoImport = '';
  }
  if (autoImport === true) {
    autoImport = 'power-assert';
  }

  // staticTruthCheck
  if (staticTruthCheck === undefined) {
    ({ staticTruthCheck } = defaultConfig);
  }

  // assertFunctionName
  if (assertFunctionName === undefined) {
    ({ assertFunctionName } = defaultConfig);
  }

  // presets
  if (presets === undefined) {
    ({ presets } = defaultConfig);
  }
  const hooks = presets.map(preset => require(preset) as Hooks).reduce(
    (accHooks, moreHooks) => ({
      assertionPostProcessors: [
        ...accHooks.assertionPostProcessors,
        ...moreHooks.assertionPostProcessors,
      ],
    }),
    { assertionPostProcessors: [] },
  );

  return {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    hooks,
  };
};
