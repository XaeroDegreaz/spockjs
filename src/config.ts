export interface Config {
  powerAssert: boolean;
}

const defaultConfig: Config = {
  powerAssert: true,
};

export default defaultConfig;
