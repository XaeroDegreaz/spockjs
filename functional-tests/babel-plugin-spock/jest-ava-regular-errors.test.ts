import { transform } from '@babel/core';
import assert, { AssertionError } from 'power-assert';

import { Config, minimalConfig } from '@spockjs/config';

import autoImportDisabled from '@spockjs/assertion-post-processor-regular-errors/src/auto-import-disabled-warning';

import plugin from '@spockjs/babel-plugin-spock';

// mark implicit dependencies for jest
() => require('@spockjs/runner-jest') && require('@spockjs/runner-ava');

const originalConsoleWarn = console.warn;
let consoleWarn: jest.Mock;
beforeEach(() => (console.warn = consoleWarn = jest.fn()));
afterEach(() => (console.warn = originalConsoleWarn));

test('throws plain Errors instead of AssertionErrors', () => {
  const { code } = transform(`expect: 1 === 2;`, {
    plugins: [
      [
        plugin,
        { ...minimalConfig, presets: ['@spockjs/runner-jest'] } as Config,
      ],
    ],
  });

  expect.assertions(1);
  try {
    new Function('assert', 'AssertionError', code as string)(
      assert,
      AssertionError,
    );
  } catch (e) {
    expect: !(e instanceof AssertionError);
    expect(e).toMatchSnapshot();
  }
});

test('throws plain Errors instead of AssertionErrors with runner-ava instead of runner-jest', () => {
  const { code } = transform(`expect: 1 === 2;`, {
    plugins: [
      [
        plugin,
        { ...minimalConfig, presets: ['@spockjs/runner-ava'] } as Config,
      ],
    ],
  });

  expect.assertions(1);
  try {
    new Function('assert', 'AssertionError', code as string)(
      assert,
      AssertionError,
    );
  } catch (e) {
    expect: !(e instanceof AssertionError);
    expect(e).toMatchSnapshot();
  }
});

test('leaves other errors untouched', () => {
  class CustomError extends Error {}
  const throws = () => {
    throw new CustomError();
  };
  const { code } = transform(`expect: throws() === 1;`, {
    plugins: [
      [
        plugin,
        { ...minimalConfig, presets: ['@spockjs/runner-jest'] } as Config,
      ],
    ],
  });

  expect(() => {
    new Function('assert', 'AssertionError', 'throws', code as string)(
      assert,
      AssertionError,
      throws,
    );
  }).toThrow(CustomError);
});

describe('autoImport disabled warning', () => {
  beforeEach(() => (autoImportDisabled.warned = false));

  test('warns about the need to import AssertionError', () => {
    transform(`expect: 1 === 2;`, {
      plugins: [
        [
          plugin,
          { ...minimalConfig, presets: ['@spockjs/runner-jest'] } as Config,
        ],
      ],
    });

    expect(consoleWarn.mock.calls).toMatchSnapshot();
  });

  test('warns only once in spite of multiple traversals and assertions', () => {
    const doTransform = () =>
      transform(
        `expect: 1 === 2;
        expect: 2 === 3;`,
        {
          plugins: [
            [
              plugin,
              { ...minimalConfig, presets: ['@spockjs/runner-jest'] } as Config,
            ],
          ],
        },
      );
    doTransform();
    doTransform();

    expect(consoleWarn).toHaveBeenCalledTimes(1);
  });
});

test('does not require AssertionError in scope with autoImport enabled', () => {
  const { code } = transform(`expect: 1 === 2;`, {
    plugins: [
      [
        plugin,
        {
          ...minimalConfig,
          presets: ['@spockjs/runner-jest'],
          autoImport: true,
        } as Config,
      ],
    ],
    presets: ['@babel/preset-env'],
  });

  expect.assertions(1);
  try {
    new Function('require', code as string)(require);
  } catch (e) {
    expect: !(e instanceof AssertionError);
    expect(e).toMatchSnapshot();
  }
});

test('keeps the pretty error message from powerAssert', () => {
  const { code } = transform(`expect: 1 === 2;`, {
    plugins: [
      [
        plugin,
        {
          ...minimalConfig,
          presets: ['@spockjs/runner-jest'],
          powerAssert: true,
        } as Config,
      ],
    ],
    filename: 'test.js',
  });

  expect.assertions(1);
  try {
    new Function('assert', 'AssertionError', code as string)(
      assert,
      AssertionError,
    );
  } catch (e) {
    expect: !(e instanceof AssertionError);
    expect(e).toMatchSnapshot();
  }
});
