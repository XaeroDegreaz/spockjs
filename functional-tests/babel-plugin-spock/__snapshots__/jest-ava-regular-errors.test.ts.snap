// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoImport disabled warning warns about the need to import AssertionError 1`] = `
Array [
  Array [
    "Generating catch clauses to intercept AssertionErrors with autoImport disabled. The AssertionError class must be imported and available in scope as \`AssertionError\`",
  ],
]
`;

exports[`does not require AssertionError in scope with autoImport enabled 1`] = `[Error: false == true]`;

exports[`keeps the pretty error message from powerAssert 1`] = `
[Error:   # test.js:1
  
  assert(1 === 2)
           |     
           false 
  
  [number] 2
  => 2
  [number] 1
  => 1
  ]
`;

exports[`throws plain Errors instead of AssertionErrors 1`] = `[Error: false == true]`;

exports[`throws plain Errors instead of AssertionErrors with runner-ava instead of runner-jest 1`] = `[Error: false == true]`;
